oracle: add oracle web3 endpoint pool

jordipainan opened the pull request:
If an Oracle that is subscribed to one or more
ethereum smart contract events
have any problem with the endpoint that is
connected with, after some retries, another
web3 endpoint will be used.

If there is not any web3 endpoint working
the application will terminate.

When a new connection is created the ethevents
service will be restarted as well as the
PrintInfo go routine.

A context.WithCancel is being used to terminate
all the goroutines created by the ethereum events
service and the contracts will be resolved on
the ENS an instanciated again. That way if any
event is lost during the time that there
are connection issues, will be catch again
on restart.

jordipainan committed:
oracle: add oracle web3 endpoint pool

If an Oracle that is subscribed to one or more
ethereum smart contract events
have any problem with the endpoint that is
connected with, after some retries, another
web3 endpoint will be used.

If there is not any web3 endpoint working
the application will terminate.

When a new connection is created the ethevents
service will be restarted as well as the
PrintInfo go routine.

A context.WithCancel is being used to terminate
all the goroutines created by the ethereum events
service and the contracts will be resolved on
the ENS an instanciated again. That way if any
event is lost during the time that there
are connection issues, will be catch again
on restart.

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -543,6 +543,28 @@ func main() {
+				}
+			}
+		} else {
+			log.Errorf(`No web3 endpoint found, oracle needs at least
```
That should probably be a Fatal... or if its an Error do something

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -543,6 +543,28 @@ func main() {
+			one valid web3 endpoint to handle Ethereum contract events`)
+		}
+
+		log.Debugf("Using %+v web3 endpoints", w3uris)
```
capital letter

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -543,6 +543,28 @@ func main() {
 		}
 
+		w3uris := []string{}
+		if len(globalCfg.W3Config.W3External) > 0 {
```
I'd revert that logic... its a bit messy.

Remove the `if len(globalCfg.W3Config.W3External) > 0 {` and also the else and just add after the for

`if len(w3uris) == 0 { log.Fatal("no valid web3 endpoint found") }`

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -543,6 +543,27 @@ func main() {
+			log.Fatal(`no web3 endpoint found, oracle needs at least
+			one valid web3 endpoint to handle Ethereum contract events`)
+		}
+		log.Debugf("using %+v web3 endpoints", w3uris)
```
this should be info IMO

jordipainan replied:
changed to info :)

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -559,56 +580,45 @@ func main() {
+				initBlock = new(int64)
+				chainSpecs, err := ethchain.SpecsFor(globalCfg.W3Config.ChainType)
+				if err != nil {
+					log.Warn("cannot get chain block to start looking for events, using 0")
```
I think initBlock should not be a pointer, what's the reason? 

Also I'm not sure we should uset initBlock=0 if error. I think its safer to do a Fatal() and print the error instead

jordipainan replied:
yes agree, at some point we discussed to not fatal in this kind of cases, that is why most of this errors are handled with a warning. In any case I also prefer to not use initBlock=0

p4u replied:
If an error is something that does not allow the execution, it should be a Fatal. And still, why initBlock is a pointer?

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -559,56 +580,45 @@ func main() {
+				if err != nil {
+					log.Warn("cannot get chain block to start looking for events, using 0")
+					*initBlock = 0
+				} else {
```
No need for this else if err executes Fatal()

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -559,56 +580,45 @@ func main() {
+
+			whiteListedAddr := []string{}
+			for _, addr := range globalCfg.VochainConfig.EthereumWhiteListAddrs {
+				if ethcommon.IsHexAddress(addr) {
```
If `addr` is not Hex, shouldn't we at least print a Warning?

jordipainan replied:
yes, added

p4u commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -632,7 +642,7 @@ func main() {
 				log.Fatal(err)
 			}
 			if err := apior.EnableERC20(globalCfg.W3Config.ChainType,
```
I think breaking lines here is not needed

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -63,6 +63,7 @@ type EthereumEvents struct {
 	ContractsAddress []common.Address
 	// ContractsInfo holds useful info for working with the desired contracts
 	ContractsInfo map[string]*ethereumhandler.EthereumContract
+	StartBlock    *int64
```
why a pointer?

p4u replied:
If you thinking about saving memory, take into account you are actually doing the opposite sice a pointer will require 2xint64 (one for the address and another for the value)

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
+	go func(chan bool) {
+		for {
+			time.Sleep(time.Second * 5)
+			blockTctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout*2)
```
Why EthereumReadTimeout x2 ?

jordipainan replied:
copy paste err, changed

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
+	go func(chan bool) {
+		for {
+			time.Sleep(time.Second * 5)
+			blockTctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout*2)
```
Just call it ctx (more standard)

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
+			if _, err := ev.VotingHandle.EthereumClient.BlockByNumber(blockTctx, nil); err != nil {
+				log.Warnf("cannot check ethereum connection while running the event processor, %s", err)
+				connCh <- false
+				cancel()
```
cancel() should also be called if not error

jordipainan replied:
moved and added a defer that will be executed in the very end

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
+		for {
+			time.Sleep(time.Second * 5)
+			blockTctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout*2)
+			if _, err := ev.VotingHandle.EthereumClient.BlockByNumber(blockTctx, nil); err != nil {
```
Shouldn't the block number be saved and check that it is actually being increased? Else we won't detect a frozen web3 endpoint

jordipainan replied:
there is an else clause that will be executed when restarting the subscription (100 blocks), I think is enough for this purpose. What do you think?

jordipainan replied:
will tackle on the next PR, added a better approach :)

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
 	}
 
+	// check eth connection
+	connCh := make(chan bool, 1)
```
This channel should have a more descriptive name, such as "connFailure"

jordipainan replied:
changed :)

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -377,20 +400,28 @@ func (ep *EventProcessor) next() *ethtypes.Log {
+					ev.EventProcessor.del(&event)
+				} else {
+					if event.BlockNumber == 0 {
+						log.Fatalf("this event %+v must not be arribe here (BOMB)", event)
```
what's the meaning of this message? I don't get it. Can you rephrase?

jordipainan replied:
just a sanity check cause I do not expect having block number 0 on event logs. Changed the message for more clarity

p4u commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
+				log.Warnf("could not connect to %s endpoint, trying the next one ...", endpoint)
+				break
+			}
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
```
Use inline if

p4u commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
+		maxtries := 5
+		for {
+			if maxtries == 0 {
+				log.Warnf("could not connect to %s endpoint, trying the next one ...", endpoint)
```
"one..." (or even better, remove the "..." since we are not using that in general)

p4u commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
+			}
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
+			if err != nil || eh.EthereumRPC == nil {
+				log.Warnf("cannot create an ethereum rpc connection with : (%s), trying again ...", err)
```
`"with %s: (%v)", endpoint, err)`

p4u commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
+			if err != nil || eh.EthereumRPC == nil {
+				log.Warnf("cannot create an ethereum rpc connection with : (%s), trying again ...", err)
+				time.Sleep(time.Second * 5)
```
Why sleeping 5 seconds? Seems too much

jordipainan replied:
lowered to 3

p4u commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -163,7 +166,10 @@ func (eh *EthereumHandler) PrintInfo(ctx context.Context, seconds time.Duration)
-			log.Warn(err)
+			log.Warnf("error getting ethereum info: %s", err)
+			if err := eh.Connect(eh.Endpoints); err != nil {
+				log.Warnf("cannot connect to any web3 endpoint: %v", err)
```
Shouldn't this Fatal?

jordipainan replied:
yes assuming last comment made on this, changed to a fatal

p4u commented on service/ethevents.go:
```diff
@@ -40,20 +41,37 @@ func EthEvents(
-		ev.SubscribeEthereumEventLogs(ctx, startBlock)
-	}()
+
+	initialized := new(bool)
```
Why a pointer?

jordipainan replied:
no need, only one goroutine use the var, changed and no longer using a pointer

p4u commented on service/ethevents.go:
```diff
@@ -40,20 +41,37 @@ func EthEvents(
+			}
+			ctx, cancel := context.WithCancel(ctx)
+			if *initialized {
+				if ev.VotingHandle, err = ethereumhandler.NewEthereumHandler(ev.ContractsInfo, specs.NetworkSource, w3uris); err != nil {
```
This line is 126 chars

p4u commented on service/ethevents.go:
```diff
@@ -40,20 +41,37 @@ func EthEvents(
+					log.Fatalf("cannot restart ethereum events, cannot create ethereum handler: %v", err)
+				}
+			}
+			*initialized = true
```
I don't understand the logic behind this routine. Why a for loop? "initialized" is set to true on each iteration, why? Also why ev.VotingHandle is inside the for loop?

jordipainan replied:
kind of a workaround because new EthEvents already invokes new ethereum handler and I wanted to avoid this. Solving properly requires some refactor on the new EthEvents function. This is out of the scope of this commit.
WIll add a TODO explaining more about this for adding clarity and i'll make another PR improving this.

jordipainan replied:
will be better addressed on the next pr

mvdan commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
+				continue
+			}
+			// if RPC connection established, create an ethereum client using the RPC client
+			eh.EthereumClient = ethclient.NewClient(eh.EthereumRPC)
```
are you sure that the EthereumClient field isn't racy? it's used from other goroutines as far as I can tell.

jordipainan replied:
yes it is, adding a TODO for having 2 different connections:
- One for the subscription
- One for the ethereum sync status

mvdan replied:
Hm, OK. Seems like a pretty big TODO, but I guess "it usually works" is better than delaying this PR.

mvdan commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
+			return
 		case event := <-logs:
 			ev.EventProcessor.Events <- event
+		case <-connCh:
```
we have two identical cases in this select now; why do we have two channels instead of one?

jordipainan replied:
discussed privately and solved

jordipainan replied:
btw, the fallthrough clause cannot be applied here afaik. The compiler complains about it and I only solved it with a weird if clause.

mvdan replied:
You're right, I forgot that fallthroughs can't be used in selects. I guess not worth doing anything more magic here, but we could at least put both cases next to each other.

mvdan commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -545,6 +545,27 @@ func main() {
+
+		if len(w3uris) == 0 {
+			log.Fatal(`no web3 endpoint found, oracle needs at least
+			one valid web3 endpoint to handle Ethereum contract events`)
```
FYI this log line will have a newline and lots of whitespace. I'd just join it as a single line, but if you really want to split it:

```
log.Fatal("foo bar " + 
    "baz")
```

mvdan commented on cmd/dvotenode/dvotenode.go:
```diff
@@ -561,56 +582,48 @@ func main() {
+				vnode,
+				evh,
+				sc,
+				whiteListedAddr); err != nil {
```
I really think we're going way, way overkill with avoiding long lines. You're literally writing multiple lines which are >80% whitespace :) This entire call can very easily fit in 2-3 lines.

mvdan commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +291,38 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
 		select {
-		case err := <-sub.Err():
+		case <-sub.Err():
 			// TODO: @jordipainan do not fatal here, handle error on event subscription channel
```
I think this TODO is outdated now

mvdan reviewed (approved):
SGTM, just a few minor nits

jordipainan committed:
ethevents: decouple EthereumHandler from EthEvents
and simplify the event subscription connection logic.

EthEvents was calling NewEthereumHandler on
the object creation itself and because of this the
function usage was less flexible.
Now the EthEvents EthereumHandle is added after the
object instanciation thus making the EthEvents
subscription go-routine simpler.

The startBlock attribute was also deleted in favor
of EthereumLastKnownBlock which is an atomic value
shared among different go routines.

The former follows the subscription connection
logic simplification.
The SubscribeEthereumEventLogs function had a
weird/complex logic for reading some past events
during the initial connection and now also with
the reconnection thanks to the web3 endpoints pool,
so the new logic handles the subscription as follows:
- If subscribeOnly flag == true:
  - Get latest block and start from latestBlock - readBlocksPast
  - If latest block cannot be fetch -> Fatal
- If subscribeOnly flag = false
  - Get start block from the specs and use it as latestBlock
  - If start block does not exist for the spec -> Fatal

In both cases, once the initialization is done, the
EthereumLastKnownBlock attribute will keep track of
the latest known block of an specific web3 client and
if at some point the connection is dropped and the
subscription is initialized again, the start block for the
subscription is going to be the latest known block which is
updated every 5 seconds minus the readBlocks past for adding
some extra security.

The last change made for having the subscription connection
mechanism simplified is to share the same context with
cancel with the PrintInfo function because they share the
same web3 client. From now on the PrintInfo blocking function
is terminated if the subscription connection is not working
and reinitialized in the same goroutine that keeps track
of the ethereum events subscription.

p4u commented on ethereum/ethevents/events.go:
```diff
@@ -290,13 +291,37 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
+				connFailure <- false
+				return
+			}
+			ev.EthereumLastKnownBlock = block.Number().Int64()
```
This var is never used and also it looks like racy, I think sync.Atomic should be used here

jordipainan replied:
yep, so finally adding the commit I mentioned regarding this case here. And yes, is used on the resubscribe routine

mvdan commented on ethereum/ethevents/events.go:
```diff
@@ -172,155 +161,81 @@ func (ev *EthereumEvents) AddEventHandler(handler EventHandler) {
-		lastBlock = blk.Number().Int64()
-		break
+	var err error
+	uLastBlockNumber, err := ev.VotingHandle.EthereumClient.BlockNumber(tctx)
```
you only need to do this remote ethereum request if fromBlock==0, right? I'd move it inside the "if".

mvdan commented on ethereum/ethevents/events.go:
```diff
@@ -172,155 +161,81 @@ func (ev *EthereumEvents) AddEventHandler(handler EventHandler) {
+	lastBlockNumber := int64(uLastBlockNumber)
+	fromBlock := ev.EthereumLastKnownBlock.Load().(int64)
+	if fromBlock == 0 {
+		ev.EthereumLastKnownBlock.Store(lastBlockNumber)
```
note that this Store is separate from the Load, so it's not atomic. I imagine we don't mind that. If the original atomic value was 0 and two goroutines both try to update it at the same time, I imagine they'd both try to update it to the same new value. If it's possible that those two goroutines would end up with different "latest block" values, then that might be problematic.

mvdan commented on ethereum/handler/ethereumHandler.go:
```diff
@@ -161,15 +161,15 @@ func (eh *EthereumHandler) PrintInfo(ctx context.Context, seconds time.Duration)
+		if ctx.Err() != nil {
+			return
+		}
 		time.Sleep(seconds)
```
nit: do a select between ctx.Done and time.After, to stop sleeping if the context is cancelled

mvdan reviewed (commented):
Sorry for the late review. Better late than never :) Mostly just nits.
